File: backend/app/services/websocket_service.py
import socketio
from typing import Dict, Any, Optional
import logging
from fastapi import HTTPException
import json

logger = logging.getLogger(__name__)

class WebSocketService:
    def __init__(self):
        self.sio = socketio.AsyncServer(
            async_mode='asgi',
            cors_allowed_origins="*",
            logger=False,
            engineio_logger=False
        )
        self.app = socketio.ASGIApp(self.sio)
        self.active_conversations: Dict[str, Dict[str, Any]] = {}
        self.user_sessions: Dict[str, str] = {}  # session_id -> user_id mapping
        
        self._setup_event_handlers()
    
    def _setup_event_handlers(self):
        @self.sio.event
        async def connect(sid, environ, auth):
            # Extract token from auth if provided
            if auth and 'token' in auth:
                # Here you would validate the token and get user_id
                # For now, we'll store the session
                self.user_sessions[sid] = auth.get('user_id', sid)
            return True
        
        @self.sio.event
        async def disconnect(sid):
            # Clean up any active conversations for this session
            for conv_id, conv_data in list(self.active_conversations.items()):
                if conv_data.get('sid') == sid:
                    await self.stop_conversation(conv_id)
            # Remove user session
            self.user_sessions.pop(sid, None)
        
        @self.sio.event
        async def join_conversation(sid, data):
            """Join a conversation room"""
            conversation_id = data.get('conversation_id')
            if conversation_id:
                self.sio.enter_room(sid, conversation_id)
                await self.sio.emit('joined_conversation', {
                    'conversation_id': conversation_id,
                    'status': 'joined'
                }, room=sid)
        
        @self.sio.event
        async def leave_conversation(sid, data):
            """Leave a conversation room"""
            conversation_id = data.get('conversation_id')
            if conversation_id:
                self.sio.leave_room(sid, conversation_id)
                await self.sio.emit('left_conversation', {
                    'conversation_id': conversation_id,
                    'status': 'left'
                }, room=sid)
        
        @self.sio.event
        async def start_conversation(sid, data):
            """Start a new multi-agent conversation"""
            try:
                conversation_id = data.get('conversation_id')
                agent_ids = data.get('agent_ids', [])
                initial_message = data.get('initial_message', '')
                max_turns = data.get('max_turns', 10)
                user_id = data.get('user_id') or self.user_sessions.get(sid)
                file_attachments = data.get('file_attachments')
                project_id = data.get('project_id')
                
                # Store conversation data
                self.active_conversations[conversation_id] = {
                    'sid': sid,
                    'agent_ids': agent_ids,
                    'user_id': user_id,
                    'is_active': True
                }
                
                # Join the conversation room
                self.sio.enter_room(sid, conversation_id)
                
                # Emit conversation started event
                await self.sio.emit('conversation_started', {
                    'conversation_id': conversation_id,
                    'agent_ids': agent_ids,
                    'status': 'started'
                }, room=conversation_id)
                
            except Exception as e:
                await self.sio.emit('conversation_error', {
                    'conversation_id': conversation_id,
                    'error': str(e)
                }, room=sid)
        
        @self.sio.event
        async def stop_conversation(sid, data):
            """Stop an active conversation"""
            conversation_id = data.get('conversation_id')
            await self.stop_conversation(conversation_id)
    
    async def stop_conversation(self, conversation_id: str):
        """Stop a conversation and notify all participants"""
        if conversation_id in self.active_conversations:
            self.active_conversations[conversation_id]['is_active'] = False
            
            await self.sio.emit('conversation_stopped', {
                'conversation_id': conversation_id,
                'status': 'stopped'
            }, room=conversation_id)
            
            # Clean up
            del self.active_conversations[conversation_id]
    
    async def emit_typing_indicator(self, conversation_id: str, agent_id: str, agent_name: str):
        """Emit typing indicator for an agent"""
        await self.sio.emit('agent_typing', {
            'type': 'typing',
            'conversation_id': conversation_id,
            'agent_id': agent_id,
            'speaker': agent_name,
            'message': '...'
        }, room=conversation_id)
    
    async def emit_agent_message(self, conversation_id: str, agent_id: str, agent_name: str, 
                                message: str, timestamp: float):
        """Emit a message from an agent"""
        await self.sio.emit('agent_message', {
            'type': 'message',
            'conversation_id': conversation_id,
            'agent_id': agent_id,
            'speaker': agent_name,
            'message': message,
            'timestamp': timestamp
        }, room=conversation_id)
    
    async def emit_conversation_complete(self, conversation_id: str):
        """Emit conversation complete event"""
        await self.sio.emit('conversation_complete', {
            'type': 'complete',
            'conversation_id': conversation_id,
            'status': 'completed'
        }, room=conversation_id)
        
        # Clean up conversation
        if conversation_id in self.active_conversations:
            del self.active_conversations[conversation_id]
    
    async def emit_error(self, conversation_id: str, error_message: str):
        """Emit error event"""
        await self.sio.emit('conversation_error', {
            'type': 'error',
            'conversation_id': conversation_id,
            'error': error_message
        }, room=conversation_id)
    
    def get_socketio_app(self):
        """Get the Socket.IO ASGI app"""
        return self.app
    
    def get_sio_server(self):
        """Get the Socket.IO server instance"""
        return self.sio

# Create a singleton instance
websocket_service = WebSocketService()

File: backend/src/services/websocket/websocket.service.ts
import { Server as SocketIOServer } from 'socket.io';
import { Server as HTTPServer } from 'http';

interface ConversationData {
  sid: string;
  agentIds: string[];
  userId: string;
  isActive: boolean;
  databaseId?: string | null;
}

interface SocketData {
  userId?: string;
}

type SocketServer = SocketIOServer<any, any, any, SocketData>;

export class WebSocketService {
  private io: SocketServer | null = null;
  private activeConversations: Map<string, ConversationData> = new Map();
  private userSessions: Map<string, string> = new Map();

  constructor() {}

  initializeSocketIO(server: HTTPServer): SocketServer {
    this.io = new SocketIOServer(server, {
      cors: {
        origin: '*',
        credentials: true,
      },
      path: '/socket.io',
    });

    this.setupEventHandlers();
    return this.io;
  }

  private setupEventHandlers(): void {
    if (!this.io) return;

    this.io.on('connection', socket => {
      const auth = socket.handshake.auth;
      if (auth && auth.token) {
        socket.data.userId = auth.userId || socket.id;
        if (socket.data.userId) {
          this.userSessions.set(socket.id, socket.data.userId);
        }
      }

      socket.on('disconnect', async () => {
        for (const [convId, convData] of this.activeConversations.entries()) {
          if (convData.sid === socket.id) {
            await this.stopConversation(convId);
          }
        }

        this.userSessions.delete(socket.id);
      });

      socket.on('join_conversation', (data: { conversation_id: string }) => {
        const { conversation_id } = data;
        if (conversation_id) {
          socket.join(conversation_id);
          socket.emit('joined_conversation', {
            conversation_id,
            status: 'joined',
          });
        }
      });

      socket.on('leave_conversation', (data: { conversation_id: string }) => {
        const { conversation_id } = data;
        if (conversation_id) {
          socket.leave(conversation_id);
          socket.emit('left_conversation', {
            conversation_id,
            status: 'left',
          });
        }
      });

      socket.on(
        'start_conversation',
        async (data: {
          conversation_id: string;
          agent_ids: string[];
          initial_message: string;
          max_turns?: number;
          user_id?: string;
          file_attachments?: any;
          project_id?: string;
        }) => {
          try {
            const { conversation_id, agent_ids, user_id } = data;

            const userId =
              user_id || socket.data.userId || this.userSessions.get(socket.id);

            this.activeConversations.set(conversation_id, {
              sid: socket.id,
              agentIds: agent_ids,
              userId: userId || socket.id,
              isActive: true,
            });

            socket.join(conversation_id);

            this.io?.to(conversation_id).emit('conversation_started', {
              conversation_id,
              agent_ids,
              status: 'started',
            });
          } catch (error) {
            socket.emit('conversation_error', {
              conversation_id: data.conversation_id,
              error: error instanceof Error ? error.message : 'Unknown error',
            });
          }
        }
      );

      socket.on(
        'stop_conversation',
        async (data: { conversation_id: string }) => {
          await this.stopConversation(data.conversation_id);
        }
      );
    });
  }

  async stopConversation(conversationId: string): Promise<void> {
    const conversation = this.activeConversations.get(conversationId);
    if (conversation) {
      conversation.isActive = false;

      this.io?.to(conversationId).emit('conversation_stopped', {
        conversation_id: conversationId,
        status: 'stopped',
      });

      this.activeConversations.delete(conversationId);
    }
  }

  async emitTypingIndicator(
    conversationId: string,
    agentId: string,
    agentName: string
  ): Promise<void> {
    this.io?.to(conversationId).emit('agent_typing', {
      type: 'typing',
      conversation_id: conversationId,
      agent_id: agentId,
      speaker: agentName,
      message: '...',
    });
  }

  async emitAgentMessage(
    conversationId: string,
    agentId: string,
    agentName: string,
    message: string,
    timestamp: number
  ): Promise<void> {
    this.io?.to(conversationId).emit('agent_message', {
      type: 'message',
      conversation_id: conversationId,
      agent_id: agentId,
      speaker: agentName,
      message,
      timestamp,
    });
  }

  async emitConversationComplete(conversationId: string): Promise<void> {
    this.io?.to(conversationId).emit('conversation_complete', {
      type: 'complete',
      conversation_id: conversationId,
      status: 'completed',
    });

    const conversation = this.activeConversations.get(conversationId);
    if (conversation) {
      conversation.isActive = false;
    }
  }

  async emitError(conversationId: string, errorMessage: string): Promise<void> {
    this.io?.to(conversationId).emit('conversation_error', {
      type: 'error',
      conversation_id: conversationId,
      error: errorMessage,
    });
  }

  getSocketServer(): SocketServer | null {
    return this.io;
  }

  getActiveConversations(): Map<string, ConversationData> {
    return this.activeConversations;
  }

  isConversationActive(conversationId: string): boolean {
    const conversation = this.activeConversations.get(conversationId);
    return conversation?.isActive || false;
  }

  async close(): Promise<void> {
    if (this.io) {
      const sockets = await this.io.fetchSockets();
      for (const socket of sockets) {
        socket.disconnect(true);
      }

      this.io.close();
      this.io = null;

      this.activeConversations.clear();
      this.userSessions.clear();
    }
  }
}

export const websocketService = new WebSocketService();
